python/adelphi/bin/adelphi

```diff
@@ -16,72 +39,119 @@ log = logging.getLogger('adelphi')
-# application happy
-def cluster_fn(export_fn, keyspace_names, options, cluster, keyspace_validation_fn = None):
     metadata = cluster.metadata
-    keyspace_objects = build_keyspace_objects(keyspace_names, metadata)
-    log.info("Exporting schema for keyspaces %s", ','.join((k.name for k in keyspace_objects)))
+    keyspace_objs = build_keyspace_objects(keyspace_names, metadata)
+
+    return [anonymize_keyspace(ks) for ks in keyspace_objs] if ctx.obj['anonymize'] else keyspace_objs
+
 
-    if len(keyspace_objects) == 0:
-        log.info("No keyspace selected")
+def validate_keyspaces(keyspace_objs):
+    log.info("Exporting schema for keyspaces %s", ','.join((k.name for k in keyspace_objs)))
+
+    if len(keyspace_objs) == 0:
+        log.info("No keyspaces selected")
         exit(1)
-    if keyspace_validation_fn:
-        keyspace_validation_fn(keyspace_objects)
 
-    print(export_fn(keyspace_objects, options))
 
+def export_keyspaces(ctx, export_fn, keyspace_objs):
+    options = {'rf': ctx.obj['rf']}
+    output_dir = ctx.obj["output-dir"]
 
-def build_options(ctx):
-    return {'rf': ctx.obj['rf'], 'anonymize': ctx.obj['anonymize']}
+    if output_dir is None:
+        print(export_fn(keyspace_objs, options))
+    else:
+        for keyspace_obj in keyspace_objs:
+            with open(os.path.join(output_dir, keyspace_obj.name),'w') as output_file:
+                log.info("Writing schema for keyspace " + keyspace_obj.name + " to file " + output_file.name)
+                output_file.write(export_fn([keyspace_obj], options))
 
 
+# ============================ Command implementations ============================
 @export.command()
 @click.pass_context
 def export_cql(ctx):
-    my_fn = partial(cluster_fn, export_cql_schema, ctx.obj['keyspace_names'], build_options(ctx))
-    with_cluster(my_fn, **({k:ctx.obj[k] for k in ["hosts","port","username","password"]}))
+    partial_fn = partial(get_keyspace_objects, ctx)
+    keyspace_objs = with_cluster(partial_fn, **({k:ctx.obj[k] for k in ["hosts","port","username","password"]}))
+    validate_keyspaces(keyspace_objs)
+
+    export_keyspaces(ctx, export_cql_schema, keyspace_objs)
 
 
 @export.command()
 @click.pass_context
 def export_gemini(ctx):
-    def my_export_fn(*args):
-        return to_string(export_gemini_schema(*args))
+    partial_fn = partial(get_keyspace_objects, ctx)
+    keyspace_objs = with_cluster(partial_fn, **({k:ctx.obj[k] for k in ["hosts","port","username","password"]}))
+    validate_keyspaces(keyspace_objs)
+
+    if len(keyspace_objs) > 1:
+        log.error("Gemini schema doesn't support multiple keyspaces.")
+        exit(2)
+
+    # My kingdom for decent function composition :(
+    export_keyspaces(ctx, lambda x,y: format_gemini_schema(export_gemini_schema(x,y)), keyspace_objs)
+
+
+@export.command()
+@click.option('--token', help='Access token for schema storage on Github')
+@click.pass_context
+def github(ctx, token):
```

Comment:
wdyt about making the command name `submit`? we may move away from github for the schemas repo in the future, but the command would remain the same

Reply:
Funny you should ask about this... I was mulling this very question the other day. :)

"github" is a terrible name, so that definitely shouldn't be the answer long-term, as it binds us to a specific type of repo and/or submission process.  I finally settled on the name "contribute" as being the most generic.  What do you think of that?  I could live with "submit" I suppose but I really like the explicit suggestion of submitting to the project.

Whaddya think?

Reply:
"contribute" sounds good and precise to me 👍